tiledb/api/src/array/mod.rs

```diff
@@ -661,21 +658,111 @@ impl Array {
     /// coordinate values of each dimension which have been populated.
     ///
     /// This is the union of all the non-empty domains of all array fragments.
-    pub fn nonempty_domain(&self) -> TileDBResult<Option<Vec<TypedRange>>> {
+    pub fn nonempty_domain(&self) -> TileDBResult<Option<TypedNonEmptyDomain>> {
         // note to devs: calling `tiledb_array_get_non_empty_domain`
         // looks like a huge pain, if this is ever a performance bottleneck
         // then maybe we can look into it
         (0..self.schema()?.domain()?.ndim()?)
             .map(|d| self.dimension_nonempty_domain(d))
-            .collect::<TileDBResult<Option<Vec<TypedRange>>>>()
+            .collect::<TileDBResult<Option<TypedNonEmptyDomain>>>()
     }
 }
 
 impl Drop for Array {
     fn drop(&mut self) {
         let c_array = *self.raw;
-        self.capi_call(|ctx| unsafe { ffi::tiledb_array_close(ctx, c_array) })
+        let mut c_is_open: i32 = out_ptr!();
+
+        self.capi_call(|ctx| unsafe {
+            ffi::tiledb_array_is_open(ctx, c_array, &mut c_is_open)
+        })
+        .expect("TileDB internal error when closing array");
+
+        // the array will not be open if the user constructs Opener and drops it without calling
+        // `Opener::open`.  This bit of al dente buccatini is mitigated by the fact that the
+        // user still never sees a non-open Array object. Maybe worth refactoring at some point
+        // nonetheless.
+        if c_is_open == 1 {
+            self.capi_call(|ctx| unsafe {
+                ffi::tiledb_array_close(ctx, c_array)
+            })
             .expect("TileDB internal error when closing array");
+        }
+    }
+}
+
+/// Holds configuration options for opening an array located at a particular URI.
+pub struct ArrayOpener {
+    array: Array,
+    /// Mode for opening the array, or `None` if re-opening.
+    mode: Option<Mode>,
+}
+
+impl ArrayOpener {
+    /// Prepares to open the array located at `uri` for operations indicated by `mode`.
+    pub fn new<S>(context: &Context, uri: S, mode: Mode) -> TileDBResult<Self>
+    where
+        S: AsRef<str>,
+    {
+        let mut array_raw: *mut ffi::tiledb_array_t = out_ptr!();
+        let c_uri = cstring!(uri.as_ref());
+
+        context.capi_call(|ctx| unsafe {
+            ffi::tiledb_array_alloc(ctx, c_uri.as_ptr(), &mut array_raw)
+        })?;
+
+        Ok(ArrayOpener {
+            array: Array {
+                context: context.clone(),
+                uri: uri.as_ref().to_owned(),
+                raw: RawArray::Owned(array_raw),
+            },
+            mode: Some(mode),
+        })
+    }
+
+    /// Configures the start timestamp for an open array.
```

Comment:
I think this phrasing is slightly wrong? These can't be set after an array is open (excluding the re-open semantics). So it should say something like "for an array about to be opened" or similar?